Problem: There is only one scheduler, meaning pdb serializes every operation.

omarkj opened the pull request:
Solution: Start a scheduler for every core, and
add a dispatches (currently LIFO). Since LMDB has
a single writer, add a global write lock for now.

This is a work on progress. I'm opening a PR now to
get feedback on my the current path.

Commit:
Problem: There is only one scheduler, meaning pdb
serializes every operation.

Solution: Start a scheduler for every core, and
add a dispatches (currently LIFO). Since LMDB has
a single writer, add a global write lock for now.

yrashk commented on src/server/server.rs:
```diff
@@ -266,15 +268,22 @@ impl Server {
             vec.append(&mut message);
-            let _ = self.sender.send(RequestMessage::ScheduleEnv(id, Vec::from(vec), self.response_sender.clone()));
+            if self.senders.len() - 1 > self.last_sender {
+                // This is a terrible way to balance load over senders. We need to move to something
```
I suppose uniformly random pick of the scheduler would work.

omarkj replied:
Agreed.

yrashk commented on src/bin/pumpkindb-doctests.rs:
```diff
@@ -23,10 +23,10 @@ use std::sync::mpsc;
 use pumpkindb::script::{RequestMessage, ResponseMessage, EnvId, Env, Scheduler};
 use pumpkindb::script::{textparser, binparser};
 use pumpkindb::pubsub;
+use pumpkindb::database;
```
Nitpick: I tried to avoid using "database" term to avoid having a database in the database (of course, we can't avoid the lmdb "database" term much, although it can be aliased on import), that's why I went for mod_storage vs mod_database.

yrashk commented on src/database.rs:
```diff
@@ -0,0 +1,32 @@
+
+impl<'a> GlobalWriteLock for Database<'a> {
+    fn try_lock(&self) -> bool {
+        match self.gwl.try_lock() {
```
I think this match can be replaced with `try_lock().is_ok()`

Commit:
Problem: Having a database in a database is
confusing.

Solution: Rename the database module to storage,
and rename the struct as well.

Commit:
Problem: Scheduler picking logic was so bad
calling it logic is unfair.

Solution: Introduce a slightly less bad scheduler
picking routine.

Commit:
Problem: Environment was being passed into the
Scheduler only to be passed into the `mod_storage`
as well as into the newly introduced `storage`
module.

Solution: Pass Environment into the Storage struct
only take remove a direct dependency between the
scheduler and lmdb env.

Commit:
Problem: Environment builder is currently in the
`main.rs` module, but should be in the `storage`
module since it’s used there.

Solution: Move the env builder into `sotarge.rs`.

yrashk commented:
I think this might be highly relevant: https://github.com/daschl/hwloc-rs — the only nitpick here is that they rely on a dynamic library (hwloc), but it'd much better if it was static...

What do you think of the whole issue of thread affinity?

http://nitschinger.at/Binding-Threads-And-Processes-to-CPUs-in-Rust/

P.S. I hope if we use it, we can amend the build procedure to make it a static dependency.
P.P.S. hwloc itself supports building static lib

yrashk commented:
To avoid further delay, however, we can split affinity issue into a separate discussion
